crypto/BRCryptoWallet.c

```diff
@@ -191,57 +205,83 @@ cryptoWalletGetBalance (BRCryptoWallet wallet) {
 
 extern BRCryptoBoolean
 cryptoWalletHasTransfer (BRCryptoWallet wallet,
-                        BRCryptoTransfer transfer) {
-    for (size_t index = 0; index < array_count(wallet->transfers); index++)
-        if (CRYPTO_TRUE == cryptoTransferEqual (transfer, wallet->transfers[index]))
-            return CRYPTO_TRUE;
-    return CRYPTO_FALSE;
+                         BRCryptoTransfer transfer) {
+    BRCryptoBoolean r = CRYPTO_FALSE;
+    pthread_mutex_lock (&wallet->lock);
+    for (size_t index = 0; index < array_count(wallet->transfers) && CRYPTO_FALSE == r; index++) {
+        r = cryptoTransferEqual (transfer, wallet->transfers[index]);
+    }
+    pthread_mutex_unlock (&wallet->lock);
+    return r;
 }
 
 private_extern BRCryptoTransfer
 cryptoWalletFindTransferAsBTC (BRCryptoWallet wallet,
                                BRTransaction *btc) {
-    for (size_t index = 0; index < array_count(wallet->transfers); index++)
-        if (CRYPTO_TRUE == cryptoTransferHasBTC (wallet->transfers[index], btc))
-            return cryptoTransferTake (wallet->transfers[index]);
-    return NULL;
+    BRCryptoTransfer transfer = NULL;
+    pthread_mutex_lock (&wallet->lock);
+    for (size_t index = 0; index < array_count(wallet->transfers) && NULL == transfer; index++) {
+        if (CRYPTO_TRUE == cryptoTransferHasBTC (wallet->transfers[index], btc)) {
+            transfer = cryptoTransferTake (wallet->transfers[index]);
+        }
+    }
+    pthread_mutex_unlock (&wallet->lock);
+    return transfer;
 }
 
 private_extern BRCryptoTransfer
 cryptoWalletFindTransferAsETH (BRCryptoWallet wallet,
                                BREthereumTransfer eth) {
-    for (size_t index = 0; index < array_count(wallet->transfers); index++)
-        if (CRYPTO_TRUE == cryptoTransferHasETH (wallet->transfers[index], eth))
-            return cryptoTransferTake (wallet->transfers[index]);
-    return NULL;
+    BRCryptoTransfer transfer = NULL;
+    pthread_mutex_lock (&wallet->lock);
+    for (size_t index = 0; index < array_count(wallet->transfers) && NULL == transfer; index++) {
+        if (CRYPTO_TRUE == cryptoTransferHasETH (wallet->transfers[index], eth)) {
+            transfer = cryptoTransferTake (wallet->transfers[index]);
+        }
+    }
+    pthread_mutex_unlock (&wallet->lock);
+    return transfer;
```

Comment:
Personally, I'd rather `break` from the `for ... if ...` rather than find that `NULL == transfer` condition...  But that is me.

Reply:
Compromise; I've switched to breaks for all but the cases where we are returning a boolean (ex: cryptoNetworkHasCurrency).